File: examples/cpu_detect/cpp/onnx_demo.cpp
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdio>
#include <string>
#include <vector>

#include <opencv2/opencv.hpp>
#include <opencv2/dnn.hpp>

namespace {

constexpr int kInputSize = 640;
constexpr float kScoreThreshold = 0.30f;
constexpr float kNmsThreshold = 0.45f;
constexpr int kMaxDet = 100;

struct LetterboxResult {
    cv::Mat blob;
    float scale = 1.0f;
    int pad_x = 0;
    int pad_y = 0;
};

struct Detection {
    float score = 0.0f;
    float x0 = 0.0f;
    float y0 = 0.0f;
    float x1 = 0.0f;
    float y1 = 0.0f;
};

LetterboxResult letterbox_rgb(const cv::Mat& bgr) {
    int w = bgr.cols;
    int h = bgr.rows;
    float scale = std::min((float)kInputSize / std::max(w, 1), (float)kInputSize / std::max(h, 1));
    int new_w = (int)std::round(w * scale);
    int new_h = (int)std::round(h * scale);
    cv::Mat resized;
    cv::resize(bgr, resized, cv::Size(new_w, new_h), 0, 0, cv::INTER_LINEAR);
    cv::Mat canvas(kInputSize, kInputSize, CV_8UC3, cv::Scalar(0, 0, 0));
    int pad_x = (kInputSize - new_w) / 2;
    int pad_y = (kInputSize - new_h) / 2;
    resized.copyTo(canvas(cv::Rect(pad_x, pad_y, new_w, new_h)));
    cv::Mat rgb;
    cv::cvtColor(canvas, rgb, cv::COLOR_BGR2RGB);
    cv::Mat blob = cv::dnn::blobFromImage(rgb, 1.0 / 255.0, cv::Size(kInputSize, kInputSize), cv::Scalar(), false, false, CV_32F);
    return {blob, scale, pad_x, pad_y};
}

float iou(const Detection& a, const Detection& b) {
    float x0 = std::max(a.x0, b.x0);
    float y0 = std::max(a.y0, b.y0);
    float x1 = std::min(a.x1, b.x1);
    float y1 = std::min(a.y1, b.y1);
    float iw = std::max(0.0f, x1 - x0);
    float ih = std::max(0.0f, y1 - y0);
    float inter = iw * ih;
    float aa = std::max(0.0f, a.x1 - a.x0) * std::max(0.0f, a.y1 - a.y0);
    float ab = std::max(0.0f, b.x1 - b.x0) * std::max(0.0f, b.y1 - b.y0);
    return inter / std::max(aa + ab - inter, 1e-6f);
}

std::vector<Detection> nms(std::vector<Detection> dets) {
    std::sort(dets.begin(), dets.end(), [](const Detection& a, const Detection& b) { return a.score > b.score; });
    std::vector<Detection> kept;
    std::vector<char> removed(dets.size(), 0);
    for (size_t i = 0; i < dets.size(); ++i) {
        if (removed[i]) continue;
        kept.push_back(dets[i]);
        if ((int)kept.size() >= kMaxDet) break;
        for (size_t j = i + 1; j < dets.size(); ++j) {
            if (!removed[j] && iou(dets[i], dets[j]) > kNmsThreshold) removed[j] = 1;
        }
    }
    return kept;
}

std::vector<Detection> decode(const cv::Mat& output, int image_w, int image_h, float scale, int pad_x, int pad_y) {
    if (output.dims != 3) {
        throw std::runtime_error("unexpected ONNX output dims");
    }
    int c = output.size[1];
    int n = output.size[2];
    if (c != 5) {
        throw std::runtime_error("expected output shape [1,5,N]");
    }
    const float* p = (const float*)output.data;
    std::vector<Detection> dets;
    for (int i = 0; i < n; ++i) {
        float score = p[4 * n + i];
        if (score < kScoreThreshold) continue;
        float cx = p[0 * n + i];
        float cy = p[1 * n + i];
        float bw = p[2 * n + i];
        float bh = p[3 * n + i];
        Detection d;
        d.score = score;
        d.x0 = (cx - bw * 0.5f - pad_x) / scale;
        d.y0 = (cy - bh * 0.5f - pad_y) / scale;
        d.x1 = (cx + bw * 0.5f - pad_x) / scale;
        d.y1 = (cy + bh * 0.5f - pad_y) / scale;
        d.x0 = std::max(0.0f, std::min(d.x0, (float)(image_w - 1)));
        d.y0 = std::max(0.0f, std::min(d.y0, (float)(image_h - 1)));
        d.x1 = std::max(0.0f, std::min(d.x1, (float)(image_w - 1)));
        d.y1 = std::max(0.0f, std::min(d.y1, (float)(image_h - 1)));
        if (d.x1 > d.x0 && d.y1 > d.y0) dets.push_back(d);
    }
    return nms(std::move(dets));
}

}  // namespace

int main(int argc, char** argv) {
    if (argc < 2) {
        std::printf("usage: %s image.jpg [output.jpg] [model.onnx]\n", argv[0]);
        return 1;
    }
    std::string image_path = argv[1];
    std::string output_path = argc >= 3 ? argv[2] : "onnx_det.jpg";
    std::string model_path = argc >= 4 ? argv[3] : "onnx/gait_detect_dynamic_slim.onnx";

    cv::Mat image = cv::imread(image_path, cv::IMREAD_COLOR);
    if (image.empty()) {
        std::printf("failed to read image: %s\n", image_path.c_str());
        return 1;
    }

    std::vector<Detection> dets;
    double ms = 0.0;
    try {
        cv::dnn::Net net = cv::dnn::readNetFromONNX(model_path);
        LetterboxResult input = letterbox_rgb(image);
        auto t0 = std::chrono::high_resolution_clock::now();
        net.setInput(input.blob);
        cv::Mat output = net.forward();
        dets = decode(output, image.cols, image.rows, input.scale, input.pad_x, input.pad_y);
        auto t1 = std::chrono::high_resolution_clock::now();
        ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    } catch (const cv::Exception& e) {
        std::fprintf(stderr, "failed to run ONNX with OpenCV DNN %s\n", CV_VERSION);
        std::fprintf(stderr, "%s\n", e.what());
        std::fprintf(stderr, "This model needs a newer OpenCV DNN build on some systems. The Python ONNX demo works with the current Python cv2 wheel.\n");
        return 2;
    }

    cv::Mat drawn = image.clone();
    for (const auto& d : dets) {
        int x = (int)std::round(d.x0);
        int y = (int)std::round(d.y0);
        int w = (int)std::round(d.x1 - d.x0);
        int h = (int)std::round(d.y1 - d.y0);
        std::printf("%.4f %d %d %d %d\n", d.score, x, y, w, h);
        cv::rectangle(drawn, cv::Rect(x, y, w, h), cv::Scalar(0, 0, 255), 2);
        char text[32];
        std::snprintf(text, sizeof(text), "%.2f", d.score);
        cv::putText(drawn, text, cv::Point(x, std::max(0, y - 4)), cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 0, 255), 1);
    }
    cv::imwrite(output_path, drawn);
    std::printf("detections=%zu time=%.2fms output=%s\n", dets.size(), ms, output_path.c_str());
    return 0;
}

File: examples/cpu_detect/python_demo/onnx_demo.py
from __future__ import annotations

import sys
import time
from pathlib import Path

import cv2
import numpy as np


MODEL_PATH = "onnx/gait_detect_dynamic_slim.onnx"
INPUT_SIZE = 640
SCORE_THRESHOLD = 0.30
NMS_THRESHOLD = 0.45
MAX_DET = 100


def letterbox_rgb(image_bgr: np.ndarray, size: int = INPUT_SIZE):
    h, w = image_bgr.shape[:2]
    scale = min(size / max(w, 1), size / max(h, 1))
    new_w = int(round(w * scale))
    new_h = int(round(h * scale))
    resized = cv2.resize(image_bgr, (new_w, new_h), interpolation=cv2.INTER_LINEAR)
    canvas = np.zeros((size, size, 3), dtype=np.uint8)
    pad_x = (size - new_w) // 2
    pad_y = (size - new_h) // 2
    canvas[pad_y : pad_y + new_h, pad_x : pad_x + new_w] = resized
    rgb = cv2.cvtColor(canvas, cv2.COLOR_BGR2RGB)
    blob = rgb.transpose(2, 0, 1).astype(np.float32)[None] / 255.0
    return blob, scale, pad_x, pad_y


def nms(boxes: np.ndarray, scores: np.ndarray, threshold: float):
    if len(boxes) == 0:
        return []
    x1, y1, x2, y2 = boxes.T
    areas = np.maximum(0.0, x2 - x1) * np.maximum(0.0, y2 - y1)
    order = scores.argsort()[::-1]
    keep = []
    while order.size:
        i = int(order[0])
        keep.append(i)
        if order.size == 1:
            break
        xx1 = np.maximum(x1[i], x1[order[1:]])
        yy1 = np.maximum(y1[i], y1[order[1:]])
        xx2 = np.minimum(x2[i], x2[order[1:]])
        yy2 = np.minimum(y2[i], y2[order[1:]])
        inter = np.maximum(0.0, xx2 - xx1) * np.maximum(0.0, yy2 - yy1)
        iou = inter / np.maximum(areas[i] + areas[order[1:]] - inter, 1e-6)
        order = order[1:][iou <= threshold]
    return keep


def decode(output: np.ndarray, image_w: int, image_h: int, scale: float, pad_x: int, pad_y: int):
    pred = output[0]
    if pred.shape[0] == 5:
        pred = pred.T
    if pred.shape[-1] != 5:
        raise RuntimeError(f"unexpected ONNX output shape: {output.shape}")

    scores = pred[:, 4].astype(np.float32)
    xywh = pred[:, :4].astype(np.float32)
    boxes = np.stack(
        [
            xywh[:, 0] - xywh[:, 2] * 0.5,
            xywh[:, 1] - xywh[:, 3] * 0.5,
            xywh[:, 0] + xywh[:, 2] * 0.5,
            xywh[:, 1] + xywh[:, 3] * 0.5,
        ],
        axis=1,
    )
    mask = scores >= SCORE_THRESHOLD
    boxes = boxes[mask]
    scores = scores[mask]
    if len(scores) == 0:
        return []

    boxes[:, 0] = (boxes[:, 0] - pad_x) / scale
    boxes[:, 1] = (boxes[:, 1] - pad_y) / scale
    boxes[:, 2] = (boxes[:, 2] - pad_x) / scale
    boxes[:, 3] = (boxes[:, 3] - pad_y) / scale
    boxes[:, [0, 2]] = np.clip(boxes[:, [0, 2]], 0, image_w - 1)
    boxes[:, [1, 3]] = np.clip(boxes[:, [1, 3]], 0, image_h - 1)
    valid = (boxes[:, 2] > boxes[:, 0]) & (boxes[:, 3] > boxes[:, 1])
    boxes = boxes[valid]
    scores = scores[valid]
    keep = nms(boxes, scores, NMS_THRESHOLD)[:MAX_DET]
    return [(float(scores[i]), *[float(v) for v in boxes[i]]) for i in keep]


def main():
    if len(sys.argv) != 2:
        print(f"usage: {sys.argv[0]} image.jpg", file=sys.stderr)
        raise SystemExit(1)
    image_path = Path(sys.argv[1])
    image = cv2.imread(str(image_path), cv2.IMREAD_COLOR)
    if image is None:
        raise FileNotFoundError(image_path)

    net = cv2.dnn.readNetFromONNX(MODEL_PATH)
    blob, scale, pad_x, pad_y = letterbox_rgb(image)
    t0 = time.perf_counter()
    net.setInput(blob)
    output = net.forward()
    dets = decode(output, image.shape[1], image.shape[0], scale, pad_x, pad_y)
    ms = (time.perf_counter() - t0) * 1000.0

    out = image.copy()
    for score, x0, y0, x1, y1 in dets:
        p0 = (int(round(x0)), int(round(y0)))
        p1 = (int(round(x1)), int(round(y1)))
        cv2.rectangle(out, p0, p1, (0, 0, 255), 2)
        cv2.putText(out, f"{score:.2f}", (p0[0], max(0, p0[1] - 4)), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 255), 1)
        print(f"{score:.4f} {p0[0]} {p0[1]} {p1[0] - p0[0]} {p1[1] - p0[1]}")
    out_path = image_path.with_name(image_path.stem + "_onnx_det.jpg")
    cv2.imwrite(str(out_path), out)
    print(f"detections={len(dets)} time={ms:.2f}ms output={out_path}")


if __name__ == "__main__":
    main()
